lib/presenter.rb

```diff
@@ -0,0 +1,47 @@
+# frozen_string_literal: true
+
+class Presenter
+  attr_reader :error, :weather_data
+
+  def initialize(name_converter, user_choice = nil)
+    @name_converter = name_converter
+    @user_choice = user_choice
+    @error = false
+  end
+
+  def city_choices
+    @name_converter.city_names
+  end
+
+  def valid_city_name
+    @error = 'Invalid city chosen' unless city_choices.include?(user_choice)
+    user_choice
+  end
+
+  def log_error(error_message)
+    @error = error_message
+  end
+
+  def format_weather(data)
+    format_temperatures(data)
+    data[:humidity] = "#{data[:humidity]}\u0025"
+    data[:date] = format_date(data[:unix_date])
+    @weather_data = data
+  end
+
+  private
+
+  attr_reader :name_converter, :user_choice
+
+  def format_temperatures(data)
+    data.each_pair do |key, value|
+      data[key] = "#{value.round}\u00B0C" if key.match(
+        /(temp)|(feels_like)|(temp_min)|(temp_max)/
+      )
+    end
+  end
+
+  def format_date(unix_date)
+    Time.at(unix_date).strftime('Today %a %e %b')
+  end
```

Comment:
shall I be testing it by stubbing the tests location somehow?

Reply:
I'm changing it to `.localtime` as you've recommended for now.
I would like to know your opinion on this though, as it does seem to present a minor bug.

Reply:
I think the easiest would be to do this on the **client side**, rather than on the server side.

Timezone management is a delicate skill to master, but a very important one as it can cause many bugs.

Doing it client side means you need to do it with JavaScript, so that it happens in the browser once it has downloaded the HTML sent by Sinatra.

### Logic on the server side

* display the time [in an HTML tag with semantic meaning, like `<time>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/time), 

* add a `datetime` attribute (best practice) with the date in UTC in a format supported by the `datetime` attribute (I recommend you to use the one called "_**a valid global date and time string**_" [in this list](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/time#valid_datetime_values))

* in the the server, you can calculate the string format required by `datetime` from the UTC in seconds that you store in the database. The value in a `datetime` attribute is for computers, not for humans, so it is recommended for it to be in UTC.

The `<time>` HTML tag also requires to have some **content**. This would be the display value or in other words, the human readable version of the date. 

* The server can send a **display date** in UTC. This is because we need to do progressive enhancement, which means the page **always** has to work without JavaScript, so if the user has disabled JavaScript in their browser, we should still display some date to the user.

* If JavaScript is enabled, we will replace that value with the local timezone value in their local browser using JavaScript.

```erb
<time id="weather-date" datetime="<%= @page.weather_data[:date] %>"><%= @page.display_date %></time>
```
### Logic on the client side

Then in JavaScript, you can grab the HTML tag by id, read the value in the `datetime` attribute, and generate the human readable display date in the local timezone of the user:

```javascript
let weatherDate = document.getElementById('weather-date')
let date = new Date(weatherDate.getAttribute(datetime))
weatherDate.textContent = date.toLocaleString()
```
[Check the documentation for the JavaScript Date object here](
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date).

Since JavaScript is run in the client machine and not in the server, it will display the time their local timezone.